bryce-task-framework.md

```diff
@@ -0,0 +1,147 @@
+So we first load the executables of the lowest indegree, then we execute them. But we have a problem. At the higher level we have a schedule of every 300 seconds. what if the execution of these 
+tasks take longer than that 300 seconds? How can we avoid this? Well one solution to this type of problem would be to have some sort of autoprovisional component that vertically scales our current machine to allow for faster execution,
+of tasks we predict will take longer. We can make estimates using the metadata of a previous run. If we are making estimates we need to estimate runtime across all machines. in other words we need an algorithm to come up with some O(N) 
+time complexity analysis we can use to predict how long a given subtask will take. If we cannot execute a task in the given timeframe the user requests, even after scaling our machines as high as we can, or if a subtask is scheudled to run in an interval that is longer
+than its parent, we need to throw some kind of error or change the status of a given task tree to failure.
+
+
+#### Worker Autoscaling
+So i talked a bit about some of the problems and ways we could address provisoning on a vertical level, but what if we have a massive amount of subtasks that have the same indegree? 
+We need some sort of Distributed system to handle the horizontal scaling of these task nodes. Since we are only concerned with execution results and the most expensive part of this process is running the tasks(AKA its compute intensive rather than data intensive) 
+we can keep all of the nodes in one region, maybe having backups of these workers in nearby locations (Austin, TX backup in nebraska). This way we dont have a lot of lag when communicating between the nodes asking if all of the nodes at a given level have reached 
+a status of success. 
+
+
+For a given tree we can have a counter.This counter stores the number of nodes of a given indegree in a task tree. Each time a node reaches success, we want to decrement this counter. we can have an expected completion time for each level of the tree based on task nodes of the same type. once the counter of a given indegree reaches zero, we get the ok to start executing the next level of nodes. We want to populate a queue of execution that each time we reach a new indegree, provisions an appropriate number of boxes to handle the tasks for that level. 
+
+
+```python
+
+# We can store the task statuses as numbers to save bytes and then have a enum or switch statement to handle returning the right status on the clientside of a task
+
+def get_status(status: int):
+	if status == 0:
+		return "CREATED"
+	elif status == 1:
+		return "ENQUEUED"
+	elif status == 2:
+		return "STARTED"
+	else:
+		return "FAILURE"
+```
+
+A given task will be enqueued in execution in celery pulling nodes from redis, then we update some elasticsearch instance for each node.
+
+```python
+@dataclass
+class TaskNode:
+	status: int
+	task_id: uuid
```

Comment:
Given that we have a billion rows+ every day, i think a uuid is going to cause you a bit of trouble.
UUIDs aren't sorted, so even though you can index them with a b-tree or a hash-index, they can be slow:

https://www.percona.com/blog/2019/11/22/uuids-are-popular-but-bad-for-performance-lets-discuss/

You can, however, increase throughput considerably using a sequentialish uuid w/ the timestamp encoded:

https://www.2ndquadrant.com/en/blog/sequential-uuid-generators/

Basically, databases really love sequential data, and performance can vary quite considerably if you use a uuid vs a plain int. 

Here I would ask myself if it's worth it to bottleneck at the database level given the amounts of writes i'm doing, by allow sequential id generation vs taking the performance hit (inserts can be slower with a uuid, and indexes are much bigger: https://www.cybertec-postgresql.com/en/int4-vs-int8-vs-uuid-vs-numeric-performance-on-bigger-joins/)